plugin/src/main/scala/com/github/sbt/avro/SbtAvro.scala

```diff
@@ -0,0 +1,245 @@
+      packageAvro / publishArtifact := false,
+    ) ++ packageTaskSettings(packageAvro, packageAvroMappings) ++ Seq(
+      packageAvro / artifact := (packageAvro / artifact).value.withType(Artifact.SourceType)
+    )
+  }
+
+  import autoImport._
+
+  def packageAvroMappings: Def.Initialize[Task[Seq[(File, String)]]] = Def.task {
+    avroSources.value.flatMap(src => (src ** AvroFilter).pair(relativeTo(src)))
+  }
+
+  override def trigger: PluginTrigger = noTrigger
+
+  override def requires: Plugins = sbt.plugins.JvmPlugin
+
+  override lazy val projectSettings: Seq[Setting[_]] = defaultSettings ++
+    inConfig(Avro)(Defaults.configSettings) ++
+    Seq(Compile, Test).flatMap(c => inConfig(c)(configScopedSettings))
+
+  private def unpack(deps: Seq[File],
+                     extractTarget: File,
+                     includeFilter: FileFilter,
+                     excludeFilter: FileFilter,
+                     streams: TaskStreams): Seq[File] = {
+    def cachedExtractDep(jar: File): Seq[File] = {
+      val cached = FileFunction.cached(
+        streams.cacheDirectory / jar.name,
+        inStyle = FilesInfo.lastModified,
+        outStyle = FilesInfo.exists
+      ) { deps =>
+        IO.createDirectory(extractTarget)
+        deps.flatMap { dep =>
+          val filter = includeFilter -- excludeFilter
+          val (avroSpecs, filtered) = IO
+            .unzip(dep, extractTarget, AvroFilter)
+            .partition(filter.accept)
+          IO.delete(filtered)
+          if (avroSpecs.nonEmpty) {
+            streams.log.info("Extracted from " + dep + avroSpecs.mkString(":\n * ", "\n * ", ""))
+          } else {
+            streams.log.info(s"No Avro specification extracted from $dep")
+          }
+          avroSpecs
+        }
+      }
+      cached(Set(jar)).toSeq
+    }
+
+    deps.flatMap(cachedExtractDep)
+  }
+
+  private def unpackDependenciesTask(key: TaskKey[Seq[File]]) = Def.task {
+    val conf = configuration.value.toConfigRef
+    val avroArtifacts = update
+      .value
+      .filter(avroDependencyIncludeFilter.value)
+      .toSeq
+      .collect { case (`conf`, _, _, file) => file }
+
+    unpack(
+      deps = avroArtifacts,
+      extractTarget = (key / target).value,
+      includeFilter = (key / includeFilter).value,
+      excludeFilter = (key / excludeFilter).value,
+      streams = (key / streams).value,
+    )
+  }
+
+
+
+  private def sourceGeneratorTask(key: TaskKey[Seq[File]]) = Def.task {
+    val out = (key / streams).value
+    val externalSrcDir = (avroUnpackDependencies / target).value
+    val srcDirs = Seq(externalSrcDir) ++ avroSources.value
+    val outDir = (key / target).value
+
+    val cachedCompile = {
+      import sbt.util.CacheStoreFactory
+      import sbt.util.CacheImplicits._
+
+      val cacheStoreFactory = CacheStoreFactory(out.cacheDirectory / "avro")
+      val lastCache = { (action: Option[Set[File]] => Set[File]) =>
```

Comment:
Do we need to include Avro compiler version in the cache key as well?

Reply:
changing any sbt setting involves to reload the project to take effect. the reloading operation invalidates all caches.